File: BusinessLogic/clsBL.vb
Imports System.IO
Imports System.Web
Imports System.Text

Public Class clsBL

    Private cls As clsFanfic

    Private txtResult As String

    Dim Story As New clsFanfic.Story

    Public Function GetChapters(ByVal link As String) As Boolean

        Dim htmldoc As String

        'Download Information from Source

        If link = "" Then
            Return False
        End If

        htmldoc = cls.GrabData(link)
        txtResult = htmldoc

        If InStr(LCase(txtResult), cls.ErrorMessage) = 0 Then

            If txtResult = "" Then
                Return False
            End If


            Story.Chapters = cls.GetChapters(htmldoc)



            htmldoc = cls.ProcessChapters( _
                                           link, _
                                           0 _
                                         )

            Story.StoryURL = link

            Story.ID = cls.GetStoryID(link)

            Story.Title = cls.GrabTitle(htmldoc)
            Story.Author = cls.GrabAuthor(htmldoc)

            Story.PublishDate = cls.GrabDate(htmldoc, "Published: ")
            Story.UpdateDate = cls.GrabDate(htmldoc, "Updated: ")

            If Story.UpdateDate = "" Then Story.UpdateDate = Story.PublishDate

            Story.ChapterCount = UBound(Story.Chapters)

            txtResult = cls.GrabBody(htmldoc)

            Return True

        Else
            Return False
        End If

    End Function

    Public Sub ProcessChapters( _
                                ByVal link As String, _
                                ByVal Start As Integer, _
                                ByVal Count As Integer, _
                                ByVal FileMask As String, _
                                Optional ByVal Category As String = "" _
                              )

        Dim htmldoc As String
        Dim idx As Integer
        Dim msg As String

        'Process Chapters from Source

        For idx = (Start - 1) To (Count - 1)

            msg = "Chapter " & _
                  (idx + 1) & _
                  " of " & _
                  Count

            frmMain.UpdateProgess(msg)

            htmldoc = cls.ProcessChapters( _
                                           link, _
                                           idx _
                                         )

            ProcessChapter( _
                            htmldoc, _
                            FileMask, _
                            idx, _
                            Count, _
                            link
                          )

        Next

    End Sub

    Public Sub ProcessChapter( _
                              ByRef htmlDoc As String, _
                              ByVal FileMask As String, _
                              ByVal chapter As Integer, _
                              Optional ByVal ChapterCount As Integer = 0, _
                              Optional ByVal link As String = "", _
                              Optional ByVal Category As String = ""
                            )

        Dim ecp1252 As Encoding = Encoding.GetEncoding(1252)
        Dim sr As StreamReader
        Dim sw As StreamWriter

        Dim Body As String
        Dim data As String = ""

        If ChapterCount = 0 Then
            ChapterCount = chapter
        End If

        Story.Title = cls.GrabTitle(htmlDoc)
        Story.Author = cls.GrabAuthor(htmlDoc)
        Story.Category = cls.GrabSeries(htmlDoc)
        Story.PublishDate = cls.GrabDate(htmlDoc, "Published: ")
        Story.UpdateDate = cls.GrabDate(htmlDoc, "Updated: ")

        If Story.UpdateDate = "" Then Story.UpdateDate = Story.PublishDate

        txtResult = htmlDoc

        If InStr(LCase(txtResult), "chapter not found") = 0 _
        And InStr(LCase(txtResult), "story not found") = 0 _
        Then

            data = cls.GrabBody(htmlDoc)

            Body = "<p></p>" & data

            txtResult = "<html><body>"
            txtResult += "<p>" & Story.Title & "</p>"
            txtResult += "<p>" & Story.Author & "</p>"

            If Story.Category <> "" Then
                txtResult += "<p>" & Story.Category & "</p>"
            End If

            txtResult += cls.WriteDate( _
                                        Story.PublishDate, _
                                        Story.UpdateDate, _
                                        chapter, _
                                        ChapterCount _
                                      )
            txtResult += "<p>----------------------------------</p>"
            txtResult += Body
        Else
            txtResult = "<p>Error writing file</p>"
            txtResult += "<p>Try downloading the file from " & _
                        "<a href=" & Chr(34) & link & (chapter + 1) & "/" & _
                        Chr(34) & ">here</a> in a regular browser</p>"
            txtResult += "<p>DOWNLOAD ERROR</p>"
        End If

        txtResult = HttpUtility.HtmlDecode(txtResult)

        sr = New StreamReader(StringToStream(txtResult))

        FileMask = Replace(FileMask, "-", "")

        If IsNumeric(Mid(FileMask, Len(FileMask), 1)) Then
            FileMask += "-"
        End If

        sw = New StreamWriter(Environment.GetFolderPath( _
                               Environment.SpecialFolder.Desktop) _
                               & "\" & FileMask & _
                               Format(chapter, "0#") & ".htm", _
                               False, ecp1252)
        sw.Write(sr.ReadToEnd)
        sr.Close()
        sw.Close()
        sr = Nothing
        sw = Nothing

    End Sub

#Region "Pass-Thru Functions"

    Public Function GrabStoryInfo(ByVal idx As Integer) As clsFanfic.Story

        Dim fic As clsFanfic.Story

        fic = cls.GrabStoryInfo(idx)

        Return fic

    End Function

    Public Function GrabFeed(link As String) As System.Data.DataSet

        Dim ds As DataSet

        ds = cls.GrabFeed(link)

        Return ds

    End Function

    Public Function GetStoryID(ByVal link As String) As String

        Dim id As String

        id = cls.GetStoryID(link)

        Return id

    End Function

    Public Function GetStoryInfoByID(ByVal StoryID As String) As clsFanfic.Story

        Dim fic As clsFanfic.Story

        fic = cls.GetStoryInfoByID(StoryID)

        Return fic

    End Function

    Public Function GetStoryURL(ByVal id As String) As String

        Dim link As String

        link = cls.GetStoryURL(id)

        Return link

    End Function

#End Region

#Region "Validation Routines"

    Function CheckUrl(ByRef link As String) As Boolean

        Dim host As String
        Dim ret As Boolean = False

        Dim url As URL
        url = ExtractUrl(link)

        host = url.Host

        If UBound(Split(host, ".")) = 2 Then
            host = Mid(host, InStr(host, ".") + 1)
        End If

        LoadSiteByHost(host)

        If IsNothing(cls) Then
            ret = False
        End If

        Return ret

    End Function

#End Region

#Region "Site Module Loading"

    Function LoadSiteByHost(ByVal host As String) As Boolean

        Dim ret As Boolean = False

        If Not IsNothing(cls) Then
            If host = cls.HostName Then
                Return True
            Else
                cls = Nothing
            End If
        End If

        Select Case host
            Case "fanfiction.net"
                cls = New FFNet
            Case "adultfanfiction.net"
                cls = New AFF
            Case "ficwad.com"
                cls = New FicWad
            Case "mediaminer.org"
                cls = New MM
            Case Else
                cls = Nothing
        End Select

        If Not IsNothing(cls) Then
            ret = True
        End If

        Return ret

    End Function

#End Region

#Region "Properties"

    Public ReadOnly Property Result As String
        Get
            Return txtResult
        End Get
    End Property

    Public ReadOnly Property FanFic As clsFanfic.Story
        Get
            Return Story
        End Get
    End Property

    Public ReadOnly Property Host As String
        Get
            If Not IsNothing(cls) Then
                Return cls.HostName
            End If
        End Get
    End Property

    Public ReadOnly Property Name As String
        Get
            If Not IsNothing(cls) Then
                Return cls.Name
            Else
                Return ""
            End If
        End Get
    End Property

    Public ReadOnly Property ErrorMessage As String
        Get
            If Not IsNothing(cls) Then
                Return cls.ErrorMessage
            Else
                Return ""
            End If
        End Get
    End Property

#End Region

End Class

File: FFDownload/HtmlGrabber/BusinessLogic/clsBL.cs
using System.Data;
using System.IO;
using System.Text;
using System.Web;
using Microsoft.VisualBasic;

namespace HtmlGrabber
{

    public class clsBL
    {


        private string txtResult;
        private clsFanfic.Story Story = new clsFanfic.Story();

        public string OutputPath;

        public bool GetChapters(string link)
        {

            string htmldoc;

            // Download Information from Source

            if (string.IsNullOrEmpty(link))
            {
                return false;
            }

            htmldoc = modMain.cls.GrabData(link);
            txtResult = htmldoc;

            if (Strings.InStr(Strings.LCase(txtResult), modMain.cls.ErrorMessage) == 0)
            {

                if (string.IsNullOrEmpty(txtResult))
                {
                    return false;
                }


                Story.Chapters = modMain.cls.GetChapters(htmldoc);



                htmldoc = modMain.cls.ProcessChapters(link, 0);



                Story.StoryURL = link;

                Story.ID = modMain.cls.GetStoryID(link);
                Story.Title = modMain.cls.GrabTitle(htmldoc);
                Story.Author = modMain.cls.GrabAuthor(htmldoc);
                Story.Category = modMain.cls.GrabSeries(htmldoc);

                Story.PublishDate = modMain.cls.GrabDate(htmldoc, "Published: ");
                Story.UpdateDate = modMain.cls.GrabDate(htmldoc, "Updated: ");

                if (string.IsNullOrEmpty(Story.UpdateDate))
                    Story.UpdateDate = Story.PublishDate;

                Story.ChapterCount = (Information.UBound(Story.Chapters) + 1).ToString();

                txtResult = modMain.cls.GrabBody(htmldoc);

                return true;
            }

            else
            {
                return false;
            }

        }

        public void ProcessChapters(string link, int Start, int Count, string FileMask, string Category = "")





        {

            string htmldoc;
            int idx;
            string msg;

            // Process Chapters from Source

            var loopTo = Count;
            for (idx = Start; idx <= loopTo; idx++)
            {

                msg = "Chapter " + idx + " of " + Count;



                FormManagement.frmMain.UpdateProgess(msg);

                htmldoc = modMain.cls.ProcessChapters(link, idx - 1);



                ProcessChapter(ref htmldoc, FileMask, idx, Count, link, Category);






            }

        }

        public void ProcessChapter(ref string htmlDoc, string FileMask, int chapter, int ChapterCount = 0, string link = "", string Category = "")





        {

            // Dim ecp1252 As Encoding = Encoding.GetEncoding(1252)
            StreamReader sr;
            StreamWriter sw;

            string Body;
            string data = "";

            string Path;

            if (ChapterCount == 0)
            {
                ChapterCount = chapter;
            }


            txtResult = htmlDoc;

            if (Strings.InStr(Strings.LCase(txtResult), "chapter not found") == 0 & Strings.InStr(Strings.LCase(txtResult), "story not found") == 0)

            {

                Story.Title = modMain.cls.GrabTitle(htmlDoc);
                Story.Author = modMain.cls.GrabAuthor(htmlDoc);
                Story.Category = modMain.cls.GrabSeries(htmlDoc);
                Story.PublishDate = modMain.cls.GrabDate(htmlDoc, "Published: ");
                Story.UpdateDate = modMain.cls.GrabDate(htmlDoc, "Updated: ");

                if (string.IsNullOrEmpty(Story.UpdateDate))
                    Story.UpdateDate = Story.PublishDate;

                data = modMain.cls.GrabBody(htmlDoc);

                Body = "<p></p>" + data;

                txtResult = "<html><body>";
                txtResult += "<p>" + Story.Title + "</p>";
                txtResult += "<p>" + Story.Author + "</p>";

                if (!string.IsNullOrEmpty(Story.Category))
                {
                    txtResult += "<p>" + Story.Category + "</p>";
                }

                txtResult += modMain.cls.WriteDate(Story.PublishDate, Story.UpdateDate, chapter, ChapterCount);




                txtResult += "<p>----------------------------------</p>";
                txtResult += Body;
            }
            else
            {
                txtResult = "<p>Error writing file</p>";
                txtResult += "<p>Try downloading the file from " + "<a href=" + '"' + link + chapter + "/" + '"' + ">here</a> in a regular browser</p>";

                txtResult += "<p>DOWNLOAD ERROR</p>";
            }

            txtResult = HttpUtility.HtmlDecode(txtResult);

            sr = new StreamReader(modUtility.StringToStream(txtResult));

            FileMask = Strings.Replace(FileMask, "-", "");

            if (Information.IsNumeric(Strings.Mid(FileMask, Strings.Len(FileMask), 1)))
            {
                FileMask += "-";
            }

            Path = OutputPath;

            if (!string.IsNullOrEmpty(Category))
            {
                Path += @"\";
                Path += Strings.Replace(Category, " ", "_");
            }

            if (!Directory.Exists(Path))
            {
                Directory.CreateDirectory(Path);
            }

            sw = new StreamWriter(Path + @"\" + FileMask + Strings.Format(chapter, "0#") + ".htm", false);





            sw.Write(sr.ReadToEnd());
            sr.Close();
            sw.Close();
            sr = null;
            sw = null;

        }

        public void ProcessError(string link, int chapter, string filemask, string Category = "")
        {

            var ecp1252 = Encoding.GetEncoding(1252);
            StreamReader sr;
            StreamWriter sw;

            string Path;

            string txtResult;

            txtResult = "<p>Error writing file</p>";
            txtResult += "<p>Try downloading the file from " + "<a href=" + '"' + link + chapter + '"' + ">here</a> in a regular browser</p>";

            txtResult += "<p>DOWNLOAD ERROR</p>";

            txtResult = HttpUtility.HtmlDecode(txtResult);

            sr = new StreamReader(modUtility.StringToStream(txtResult));

            filemask = Strings.Replace(filemask, "-", "");

            if (Information.IsNumeric(Strings.Mid(filemask, Strings.Len(filemask), 1)))
            {
                filemask += "-";
            }

            Path = OutputPath;

            if (!string.IsNullOrEmpty(Category))
            {
                Path += @"\";
                Path += Strings.Replace(Category, " ", "_");
            }

            if (!Directory.Exists(Path))
            {
                Directory.CreateDirectory(Path);
            }

            sw = new StreamWriter(Path + @"\" + filemask + Strings.Format(chapter, "0#") + ".htm", false, ecp1252);






            sw.Write(sr.ReadToEnd());
            sr.Close();
            sw.Close();
            sr = null;
            sw = null;

        }

        #region Pass-Thru Functions

        public clsFanfic.Story GrabStoryInfo(int idx)
        {

            clsFanfic.Story fic;

            fic = modMain.cls.GrabStoryInfo(idx);

            return fic;

        }

        public DataSet GrabFeed(string link)
        {

            DataSet ds;

            ds = modMain.cls.GrabFeed(ref link);

            return ds;

        }

        public string GetStoryID(string link)
        {

            string id;

            id = modMain.cls.GetStoryID(link);

            return id;

        }

        public clsFanfic.Story GetStoryInfoByID(string StoryID)
        {

            clsFanfic.Story fic;

            fic = modMain.cls.GetStoryInfoByID(StoryID);

            return fic;

        }

        public string GetStoryURL(string id)
        {

            string link;

            link = modMain.cls.GetStoryURL(id);

            return link;

        }

        #endregion

        #region Validation Routines

        public bool CheckUrl(ref string link)
        {

            string host;
            bool ret = false;

            URL url;
            url = URLHelper.ExtractUrl(link);

            host = url.Host;

            if (Information.UBound(Strings.Split(host, ".")) == 2)
            {
                host = Strings.Mid(host, Strings.InStr(host, ".") + 1);
            }

            ret = LoadSiteByHost(host);

            if (modMain.cls == null)
            {
                ret = false;
            }

            return ret;

        }

        #endregion

        #region Site Module Loading

        public bool LoadSiteByHost(string host)
        {

            bool ret = false;

            if (!(modMain.cls == null))
            {
                if ((host ?? "") == (modMain.cls.HostName ?? ""))
                {
                    return true;
                }
                else
                {
                    modMain.cls = null;
                }
            }

            switch (host ?? "")
            {
                case "fanfiction.net":
                    {
                        modMain.cls = new FFNet();
                        break;
                    }
                case "adult-fanfiction.org":
                    {
                        modMain.cls = new AFF();
                        break;
                    }
                case "ficwad.com":
                    {
                        modMain.cls = new FicWad();
                        break;
                    }
                case "mediaminer.org":
                    {
                        modMain.cls = new MM();
                        break;
                    }
                case "hpfanficarchive.com":
                    {
                        modMain.cls = new HPFFA();
                        break;
                    }

                default:
                    {
                        modMain.cls = null;
                        break;
                    }
            }

            if (!(modMain.cls == null))
            {
                ret = true;
            }

            return ret;

        }

        public bool LoadSiteByName(string clsname)
        {

            bool ret = true;
            string host = "";

            modMain.cls = null;

            switch (clsname ?? "")
            {
                case "FFNet":
                    {
                        modMain.cls = new FFNet();
                        break;
                    }
                case "AFF":
                    {
                        modMain.cls = new AFF();
                        break;
                    }
                case "FicWad":
                    {
                        modMain.cls = new FicWad();
                        break;
                    }
                case "MediaMiner":
                    {
                        modMain.cls = new MM();
                        break;
                    }
                case "HPFFA":
                    {
                        modMain.cls = new HPFFA();
                        break;
                    }

                default:
                    {
                        modMain.cls = null;
                        break;
                    }
            }

            if (!(modMain.cls == null))
            {
                ret = true;
            }
            else
            {
                ret = false;
            }

            return ret;

        }

        #endregion

        #region Properties

        public string Result
        {
            get
            {
                return txtResult;
            }
        }

        public clsFanfic.Story FanFic
        {
            get
            {
                return Story;
            }
        }

        public string Host
        {
            get
            {
                if (!(modMain.cls == null))
                {
                    return modMain.cls.HostName;
                }
                else
                {
                    return "";
                }
            }
        }

        public string Name
        {
            get
            {
                if (!(modMain.cls == null))
                {
                    return modMain.cls.Name;
                }
                else
                {
                    return "";
                }
            }
        }

        public string ErrorMessage
        {
            get
            {
                if (!(modMain.cls == null))
                {
                    return modMain.cls.ErrorMessage;
                }
                else
                {
                    return "";
                }
            }
        }

        #endregion

    }
}